lightningd/lightningd.c

```diff
@@ -39,39 +72,86 @@
 #include <sys/types.h>
 #include <unistd.h>
 
+/*~ The core lightning object: it's passed everywhere, and is basically a
+ * global variable.  This new_xxx pattern is something we'll see often:
+ * it allocates and initializes a new structure, using *tal*, the heirarchitcal
+ * allocator. */
 static struct lightningd *new_lightningd(const tal_t *ctx)
 {
+	/*~ tal: each allocation is a child of an existing object (or NULL,
+	 * the top-level object).  When an object is freed, all the objects
+	 * 'tallocated' off it are also freed.  In this case, freeing 'ctx'
+	 * will free 'ld'.
+	 *
+	 * It's incredibly useful for grouping object lifetimes, as we'll see.
```

Comment:
"For the technically inclined: `tal` allocations usually build a tree, and `tal_free`ing any node in the tree will result in the entire subtree rooted at that node to be freed.

Be careful though that a any cycle in the allocation graph may end up causing an infinite loop."

Also for my personal understanding: the allocations are freed in post-order right? (first children, then the parent).

Reply:
It won't loop on cycles.  Freeing is done in two phases: object is marked destroyed, destructor called, then recurse into children, then finally free(object).  So destructors are pre-order, deallocation is post-order.

It's common to want to access parents in a destructor (eg. delete from its linked list), so deallocation has to be deferred.